parser/src/transforms/constant_propagation.rs

```diff
@@ -0,0 +1,486 @@
+            ScalarExpr::SymbolAccess(sym) => {
+                let constant_value = match sym.name {
+                    // Possibly a reference to a constant declaration
+                    ResolvableIdentifier::Resolved(ref qid) => {
+                        self.global.get(qid).cloned().map(|s| (s.span(), s.item))
+                    }
+                    // Possibly a reference to a local bound to a constant
+                    ResolvableIdentifier::Local(ref id) => {
+                        self.local.get(id).cloned().map(|s| (s.span(), s.item))
+                    }
+                    // Other identifiers cannot possibly be constant
+                    _ => None,
+                };
+                if let Some((span, constant_expr)) = constant_value {
+                    match constant_expr {
+                        ConstantExpr::Scalar(value) => {
+                            assert_eq!(sym.access_type, AccessType::Default);
+                            *expr = ScalarExpr::Const(Span::new(span, value));
+                        }
+                        ConstantExpr::Vector(value) => match sym.access_type {
+                            AccessType::Index(idx) => {
+                                *expr = ScalarExpr::Const(Span::new(span, value[idx]));
+                            }
+                            ref ty => panic!(
+                                "invalid constant reference, expected scalar access, got {:?}",
+                                ty
+                            ),
+                        },
+                        ConstantExpr::Matrix(value) => match sym.access_type {
+                            AccessType::Matrix(row, col) => {
+                                *expr = ScalarExpr::Const(Span::new(span, value[row][col]));
+                            }
+                            ref ty => panic!(
+                                "invalid constant reference, expected scalar access, got {:?}",
+                                ty
+                            ),
+                        },
+                    }
+                }
+                ControlFlow::Continue(())
+            }
+            // Fold constant expressions
+            ScalarExpr::Binary(ref mut binary_expr) => {
+                let span = binary_expr.span();
+                match self.try_fold_binary_expr(binary_expr) {
+                    Ok(maybe_folded) => {
+                        if let Some(folded) = maybe_folded {
+                            *expr = ScalarExpr::Const(Span::new(span, folded));
+                        }
+                        ControlFlow::Continue(())
+                    }
+                    Err(err) => ControlFlow::Break(err),
+                }
+            }
+            // While calls cannot be constant folded, arguments can be
+            ScalarExpr::Call(ref mut call) => self.visit_mut_call(call),
```

Comment:
note: but only for pure functions

Reply:
Yes, though it's sort of a meaningless distinction here, since it would never be possible to constant fold arguments to an evaluator anyway (the arguments can't ever be constant). Semantic analysis should catch any invalid evaluator arguments so that we don't have to consider that here.

Reply:
Yeah, I know it will just do nothing (assuming things are working correctly). I guess I just like precision in comments because I feel like it helps people who are working with the code & saves time. Things that are written in comments don't have to be known or considered. It's a tiny nit, so you can ignore it if you disagree